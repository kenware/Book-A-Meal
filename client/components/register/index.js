import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import validator from 'validator';
import PropTypes from 'prop-types';
import Header from '../header/index';
import Footer from '../footer/index';
import './index.scss';
import * as actions from '../../redux/Action/action';

export class Register extends Component {
  constructor(props) {
    super(props);
    this.state = {
      reset: 'reset',
      register: '',
      email: '',
      name: '',
      username: '',
      password: '',
      vpassword: '',
      validEmail: '',
      passwordmatch: '',
      passwordmismatch: '',
      signUp: 'SignUp',
      message: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onEmail = this.onEmail.bind(this);
    this.change = this.change.bind(this);
    this.register = this.register.bind(this);
  }
  /**
   * @param  {} props if new props from redux with error arrives
   * change the state lof register button from loading icon to text
   */
  static getDerivedStateFromProps(props) {
    if (props.errorMessage.registerError) {
      return { signUp: 'SignUp' };
    }
    return null;
  }
  /**
   * remove error props in redux store if component unmount
   */
  componentWillUnmount() {
    this.props.actions.clearMessages();
  }
  /**
   * @param  {} e set state on input onchange event
   */
  onChange(e) {
    const { state } = this;
    state[e.target.name] = e.target.value;
    this.setState(state);
  }
  /**
   * @param  {} e set state on email onchange event
   * checks if email is valid
   */
  onEmail(e) {
    if (!validator.isEmail(`${e.target.value}`)) {
      const validEmail = 'Invalid Email Address';
      this.setState({ validEmail });
    } else {
      const email = e.target.value;
      const validEmail = '';
      this.setState({ email });

      this.setState({ validEmail });
    }
  }
  /**
   * @param  {} e set state on password onchange event
   * checks if password and confirm password is same
   */
  change(e) {
    const { state } = this;
    state[e.target.name] = e.target.value;
    this.setState(state);
    if (e.target.value === this.state.password) {
      const passwordmatch = 'Password Match';
      const passwordmismatch = '';
      this.setState({ passwordmatch, passwordmismatch });
    } else {
      const passwordmismatch = 'Password do not Match';
      const passwordmatch = '';
      this.setState({ passwordmismatch, passwordmatch });
    }
  }

  /**
   * @param  {} e register method
   */
  register(e) {
    e.preventDefault();
    const {
      email,
      password,
      username,
      name
    } = this.state;
    if (this.state.passwordmismatch === '' && this.state.passwordmatch !== '' && validator.isEmail(this.state.email)) {
      this.props.actions.register(name, username, email, password);
      this.setState({
        signUp: (<div><i className="fa fa-spinner fa-spin fa-2x fa-fw" aria-hidden="true" />registering...</div>)
      });
    } else {
      this.setState({ message: 'All Field Is Required' });
    }
    this.props.actions.clearMessages();
  }
  render() {
    return (
      <div>
        <Header />
        <div className="register-container" id="register-bg">
          <div className="register-col" />
          <div className="register-wrapper" style={{ height: '500px' }}>
            <h2 className="login-header">Enter your details to SignUp</h2><br />
            <h3 className="text-center danger">{this.props.errorMessage.registerError }</h3>
            <h3 className="text-center danger">{this.state.message }</h3>
            <form className="register">
              <div className="form-field">
                <label htmlFor="name">Fullname</label>
                <input onChange={this.onChange} type="text" id="name" name="name" placeholder="Eze Kevin" required />
              </div>
              <div className="form-field">
                <label htmlFor="name">Username</label>
                <input onChange={this.onChange} type="text" id="username" name="username" placeholder="Keny" required />
              </div>
              <div className="form-field">
                <label htmlFor="name">Email<br />
                  <font color="red">{this.state.validEmail} </font>
                </label>
                <input onChange={this.onEmail} type="email" name="email" placeholder="Keny" id="email" required />

              </div>
              <div className="form-field">
                <label htmlFor="password">Password</label>
                <input onChange={this.onChange} type="password" name="password" id="password" required />
              </div>
              <div className="form-field">
                <label htmlFor="password">Confirm Password<br />
                  <font color="green">{this.state.passwordmatch }</font>
                  <font color="red">{this.state.passwordmismatch }</font>
                </label>
                <input onChange={this.change} type="password" name="vpassword" id="vpassword"required />
              </div>
              <div className="form-field">
                <button type="submit" className="button lg submit" onClick={this.register}>{this.state.signUp}</button>
              </div>
              <div className="form-field">
                <label htmlFor="inputPassword3">
                  <Link to="/login">Login</Link>&nbsp;&nbsp;&nbsp;&nbsp;
                  <Link to="/resetLink">forgot password? </Link>
                </label>
              </div>
            </form>

          </div>
          <div className="register-col" />
        </div>
        <Footer />
      </div>

    );
  }
}
Register.propTypes = {
  errorMessage: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired
};

export function mapStateToProps(state) {
  return {
    errorMessage: state.errorMessage
  };
}
export function mapDispatchToProps(dispatch) {
  return { actions: bindActionCreators(actions, dispatch) };
}
export default connect(mapStateToProps, mapDispatchToProps)(Register);
