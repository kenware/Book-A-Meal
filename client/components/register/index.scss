
.register-container {
  display: flex;
  height: auto;
}

.register-col{
  flex:1;
}

.register-wrapper{
  border: dotted darkslategray;
  background-color: white;
  margin: 100px 5% 100Px 5%;
  align-items:center;
  flex:3;
  height: auto;
}

.form-field{
  display: flex;
  flex-wrap: wrap;
  padding: 10px 10% 10px 7%;
}

.form-field label{
  padding: 10px 10px 10px 10px;
  flex:1;
}

.form-field input{
  flex:2;
  padding: 10px 10px 10px 10px;
  margin-right: 10px;
  }

.form-field button{
  flex:2;
  padding: 10px 10px 10px 10px;
  margin-right: 10px;
  margin-left: 10px;
}

.text-center {
  text-align: center;
}

.danger {
  color: red;
}

#register-bg{
  background: url('../../image/image.jpg') center center fixed; 
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  opacity: 1;;
}
