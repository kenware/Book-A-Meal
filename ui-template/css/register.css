
  
  
  .register-container{
    display: flex;
    height: auto;
  }
  .register-col{
    flex:1;
  }
  .register-wrapper{
      -moz-box-shadow: 15px 5px 5px 5px #272626;
      -webkit-box-shadow: 15px 5px 5px 5px #dfd5d5;
      box-shadow: 5px 5px 5px 5px #6e6060;
      background-color: white;
      margin-top: 150px;
      align-items:center;
     flex:3;
     height: auto;
  }
  .form-field{
      display: flex;
      padding: 10px 0;
  }
  .form-field label{
   padding: 10px 10px 10px 10px;
   flex:1;
  }
  .form-field input{
    flex:2;
    padding: 10px 10px 10px 10px;
    margin-right: 10px;
   }
   .form-field button{
    flex:2;
    padding: 10px 10px 10px 10px;
    margin-right: 10px;
    margin-left: 10px;
    
  
   }