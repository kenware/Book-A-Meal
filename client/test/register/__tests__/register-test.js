import React from 'react';
import { shallow } from 'enzyme';
import { Register, mapDispatchToProps, mapStateToProps } from '../../../components/register/index';
import { props, emptyProps } from '../registerMock';

describe('Test Register Component', () => {
  it('should render correctly', () => {
    const tree = shallow(<Register {...emptyProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should respond to change event of email input and change the state of the register Component', () => {
    const tree = shallow(<Register {...emptyProps} />);
    tree.find('#email').simulate('change', { target: { name: 'email', value: 'kevin@gmail.com' } });
    expect(tree.state('email')).toEqual('kevin@gmail.com');
  });

  it('should register user onsubmit events event', () => {
    const register = jest.spyOn(Register.prototype, 'register');
    const tree = shallow(<Register {...emptyProps} />);
    tree.find('#password').simulate('change', { target: { name: 'password', value: '12345' } });
    tree.find('#vpassword').simulate('change', { target: { name: 'vpassword', value: '12345' } });
    tree.find('#email').simulate('change', { target: { name: 'email', value: 'kevin@gmail.com' } });
    tree.find('#username').simulate('change', { target: { name: 'username', value: 'kevin' } });
    const wrapper = tree.instance();
    wrapper.register({ preventDefault: jest.fn() });
    expect(register).toHaveBeenCalled();
  });

  it('should respond to lifecycle method', () => {
    const componentWillUnmount = jest.spyOn(Register.prototype, 'componentWillUnmount');
    const componentWillReceiveProps = jest.spyOn(Register.prototype, 'componentWillReceiveProps');
    const tree = shallow(<Register {...emptyProps} />);
    const wrapper = tree.instance();
    wrapper.componentWillUnmount();
    wrapper.componentWillReceiveProps({ errorMessage: { registerError: 'error' } });
    expect(componentWillReceiveProps).toHaveBeenCalled();
    expect(componentWillUnmount).toHaveBeenCalled();
  });

  it('should respond to mapStateToProps methods', () => {
    const ownProps = { match: { params: { mealId: 1 } } };
    const tree = mapStateToProps(props, ownProps);
    expect(tree).toMatchSnapshot();
  });

  it('should respond to mapDispatchToProps methods', () => {
    const tree = mapDispatchToProps(emptyProps);
    expect(tree).toMatchSnapshot();
  });
});
