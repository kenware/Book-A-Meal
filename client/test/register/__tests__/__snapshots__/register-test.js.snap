// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Register Component should render correctly 1`] = `
<div>
  <Nav
    cart={Array []}
    component=""
    onModal={[Function]}
  />
  <ToastContainer
    autoClose={8000}
    bodyClassName={null}
    className={null}
    closeButton={
      <CloseButton
        ariaLabel="close"
      />
    }
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName={null}
    rtl={false}
    style={null}
    toastClassName={null}
    transition={[Function]}
  />
  <div
    className="register-container"
    id="register-bg"
  >
    <div
      className="register-col"
    />
    <div
      className="register-wrapper"
    >
      <h2
        className="login-header"
      >
        SignUp
      </h2>
      <br />
      <h3
        className="text-center danger"
      />
      <form
        className="register"
      >
        <div
          className="form-field"
        >
          <input
            id="name"
            name="name"
            onChange={[Function]}
            placeholder="Fullname"
            required={true}
            type="text"
          />
        </div>
        <div
          className="form-field"
        >
          <input
            autoComplete="true"
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Username"
            required={true}
            type="text"
          />
        </div>
        <span />
        <div
          className="form-field"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="email"
          />
        </div>
        <div
          className="form-field"
        >
          <input
            autoComplete="true"
            id="password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
          />
        </div>
        <span />
        <div
          className="form-field"
        >
          <input
            autoComplete="true"
            id="vpassword"
            name="vpassword"
            onChange={[Function]}
            placeholder="Confirm Password"
            required={true}
            type="password"
          />
        </div>
        <div>
          <button
            className="button signup-btn"
            name="submit"
            onClick={[Function]}
            style={
              Object {
                "float": "right",
              }
            }
            type="submit"
          >
            SignUp
          </button>
        </div>
        <br />
        <div
          className="form-field"
        >
          <Link
            replace={false}
            to="/login"
          >
            Login
          </Link>
              
          <Link
            replace={false}
            to="/resetLink"
          >
            forgot password? 
          </Link>
        </div>
      </form>
    </div>
    <div
      className="register-col"
    />
  </div>
  <footer />
</div>
`;

exports[`Test Register Component should respond to mapDispatchToProps methods 1`] = `
Object {
  "actions": Object {
    "changePassword": [Function],
    "clearMessages": [Function],
    "getNotifications": [Function],
    "getUser": [Function],
    "loadErrorMessage": [Function],
    "loadNotifications": [Function],
    "loadSuccessMessage": [Function],
    "loadUser": [Function],
    "login": [Function],
    "refreshToken": [Function],
    "register": [Function],
    "resetLink": [Function],
    "updateProfile": [Function],
    "upgrade": [Function],
  },
}
`;

exports[`Test Register Component should respond to mapStateToProps methods 1`] = `
Object {
  "errorMessage": Object {
    "registerError": "Wrong credentials",
  },
}
`;
